diff --git a/build/npm/postinstall.js b/build/npm/postinstall.js
index 18ccef884a9..98356ecae31 100644
--- a/build/npm/postinstall.js
+++ b/build/npm/postinstall.js
@@ -48,6 +48,11 @@ for (let dir of dirs) {
 		continue;
 	}
 
+	if (/^remote/.test(dir) && process.env['FLATPAK_DEST']) {
+		// don't build vscode-remote for flatpak
+		continue;
+	}
+
 	if (dir === 'build/lib/watch') {
 		// node modules for watching, specific to host node version, not electron
 		yarnInstallBuildDependencies();
