diff --git a/build/npm/postinstall.js b/build/npm/postinstall.js
index 8f8b0019a7..028df6989a 100644
--- a/build/npm/postinstall.js
+++ b/build/npm/postinstall.js
@@ -33,11 +33,6 @@ function yarnInstall(location, opts) {
 
 yarnInstall('extensions'); // node modules shared by all extensions
 
-if (!(process.platform === 'win32' && (process.arch === 'arm64' || process.env['npm_config_arch'] === 'arm64'))) {
-	yarnInstall('remote'); // node modules used by vscode server
-	yarnInstall('remote/web'); // node modules used by vscode web
-}
-
 const allExtensionFolders = fs.readdirSync('extensions');
 const extensions = allExtensionFolders.filter(e => {
 	try {
