diff --git a/src/vs/platform/product/common/product.ts b/src/vs/platform/product/common/product.ts
index 537db557d87..6568d36f431 100644
--- a/src/vs/platform/product/common/product.ts
+++ b/src/vs/platform/product/common/product.ts
@@ -56,6 +56,15 @@ else {
 		});
 	}
 
+	// Set user-defined extension gallery
+	if (env['VSCODE_EXTENSIONS_GALLERY']) {
+		let gallery: IProductConfiguration['extensionsGallery'];
+		gallery = require.__$__nodeRequire(env['VSCODE_EXTENSIONS_GALLERY']);
+		Object.assign(product, {
+			extensionsGallery: gallery
+		});
+	}
+
 	Object.assign(product, {
 		version: pkg.version
 	});
