diff --git a/extensions/git/src/ipc/ipcServer.ts b/extensions/git/src/ipc/ipcServer.ts
index 8481aa4a35d..33c869da7dd 100644
--- a/extensions/git/src/ipc/ipcServer.ts
+++ b/extensions/git/src/ipc/ipcServer.ts
@@ -18,7 +18,13 @@ function getIPCHandlePath(id: string): string {
 	}
 
 	if (process.platform !== 'darwin' && process.env['XDG_RUNTIME_DIR']) {
-		return path.join(process.env['XDG_RUNTIME_DIR'] as string, `vscode-git-${id}.sock`);
+		const xdgRuntimeDir = process.env['XDG_RUNTIME_DIR'];
+		const flatpakID = process.env['FLATPAK_ID'];
+		if (flatpakID) {
+			return path.join(xdgRuntimeDir, 'app', flatpakID, `vscode-git-${id}.sock`);
+		} else {
+			return path.join(xdgRuntimeDir, `vscode-git-${id}.sock`);
+		}
 	}
 
 	return path.join(os.tmpdir(), `vscode-git-${id}.sock`);
diff --git a/src/vs/base/parts/ipc/node/ipc.net.ts b/src/vs/base/parts/ipc/node/ipc.net.ts
index e47b6c70b23..b20cbf042c6 100644
--- a/src/vs/base/parts/ipc/node/ipc.net.ts
+++ b/src/vs/base/parts/ipc/node/ipc.net.ts
@@ -765,6 +765,19 @@ const safeIpcPathLengths: { [platform: number]: number } = {
 	[Platform.Mac]: 103
 };
 
+export function getXdgSocketPath(socketName: string): string {
+	if (XDG_RUNTIME_DIR) {
+		const flatpakID = process.env['FLATPAK_ID'];
+		if (flatpakID) {
+			return join(XDG_RUNTIME_DIR, 'app', flatpakID, socketName);
+		} else {
+			return join(XDG_RUNTIME_DIR, socketName);
+		}
+	} else {
+		return join('/run', 'user', process.getuid().toString(), socketName);
+	}
+}
+
 export function createRandomIPCHandle(): string {
 	const randomSuffix = generateUuid();
 
@@ -776,7 +789,7 @@ export function createRandomIPCHandle(): string {
 	// Mac & Unix: Use socket file
 	// Unix: Prefer XDG_RUNTIME_DIR over user data path
 	const basePath = process.platform !== 'darwin' && XDG_RUNTIME_DIR ? XDG_RUNTIME_DIR : tmpdir();
-	const result = join(basePath, `vscode-ipc-${randomSuffix}.sock`);
+	const result = getXdgSocketPath(`vscode-ipc-${randomSuffix}.sock`);
 
 	// Validate length
 	validateIPCHandleLength(result);
@@ -803,7 +816,7 @@ export function createStaticIPCHandle(directoryPath: string, type: string, versi
 
 	let result: string;
 	if (process.platform !== 'darwin' && XDG_RUNTIME_DIR && !process.env['VSCODE_PORTABLE']) {
-		result = join(XDG_RUNTIME_DIR, `vscode-${scopeForSocket}-${versionForSocket}-${typeForSocket}.sock`);
+		result = getXdgSocketPath(`vscode-${scopeForSocket}-${versionForSocket}-${typeForSocket}.sock`);
 	} else {
 		result = join(directoryPath, `${versionForSocket}-${typeForSocket}.sock`);
 	}
