diff --git a/extensions/git/src/ipc/ipcServer.ts b/extensions/git/src/ipc/ipcServer.ts
index c9de357b1a6..f8830e1d5ac 100644
--- a/extensions/git/src/ipc/ipcServer.ts
+++ b/extensions/git/src/ipc/ipcServer.ts
@@ -17,8 +17,14 @@ function getIPCHandlePath(id: string): string {
 		return `\\\\.\\pipe\\vscode-git-${id}-sock`;
 	}
 
-	if (process.env['XDG_RUNTIME_DIR']) {
-		return path.join(process.env['XDG_RUNTIME_DIR'] as string, `vscode-git-${id}.sock`);
+	const xdgRuntimeDir = process.env['XDG_RUNTIME_DIR'];
+	if (xdgRuntimeDir) {
+		const flatpakID = process.env['FLATPAK_ID'];
+		if (flatpakID) {
+			return path.join(xdgRuntimeDir, 'app', flatpakID, `vscode-git-${id}.sock`);
+		} else {
+			return path.join(xdgRuntimeDir, `vscode-git-${id}.sock`);
+		}
 	}
 
 	return path.join(os.tmpdir(), `vscode-git-${id}.sock`);
diff --git a/src/vs/base/parts/ipc/node/ipc.net.ts b/src/vs/base/parts/ipc/node/ipc.net.ts
index 80873b97c1f..f62589712fc 100644
--- a/src/vs/base/parts/ipc/node/ipc.net.ts
+++ b/src/vs/base/parts/ipc/node/ipc.net.ts
@@ -764,6 +764,19 @@ const safeIpcPathLengths: { [platform: number]: number } = {
 	[Platform.Mac]: 103
 };
 
+export function getXdgSocketPath(socketName: string): string {
+	if (XDG_RUNTIME_DIR) {
+		const flatpakID = process.env['FLATPAK_ID'];
+		if (flatpakID) {
+			return join(XDG_RUNTIME_DIR, 'app', flatpakID, socketName);
+		} else {
+			return join(XDG_RUNTIME_DIR, socketName);
+		}
+	} else {
+		return join('/run', 'user', process.getuid().toString(), socketName);
+	}
+}
+
 export function createRandomIPCHandle(): string {
 	const randomSuffix = generateUuid();
 
@@ -776,7 +789,7 @@ export function createRandomIPCHandle(): string {
 	// XDG_RUNTIME_DIR over tmpDir
 	let result: string;
 	if (XDG_RUNTIME_DIR) {
-		result = join(XDG_RUNTIME_DIR, `vscode-ipc-${randomSuffix}.sock`);
+		result = getXdgSocketPath(`vscode-ipc-${randomSuffix}.sock`);
 	} else {
 		result = join(getNodeDependencies().os.tmpdir(), `vscode-ipc-${randomSuffix}.sock`);
 	}
@@ -809,7 +822,7 @@ export function createStaticIPCHandle(directoryPath: string, type: string, versi
 
 	let result: string;
 	if (XDG_RUNTIME_DIR && !process.env['VSCODE_PORTABLE']) {
-		result = join(XDG_RUNTIME_DIR, `vscode-${scopeForSocket}-${versionForSocket}-${typeForSocket}.sock`);
+		result = getXdgSocketPath(`vscode-${scopeForSocket}-${versionForSocket}-${typeForSocket}.sock`);
 	} else {
 		result = join(directoryPath, `${versionForSocket}-${typeForSocket}.sock`);
 	}
